server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java

```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
-                status.addPrimary(shardRouting.primaryShard(), shutdown);
+                status.addPrimary(shardRouting.primaryShard(), state, shutdown, includeDetails);
                 for (ShardRouting replicaShard : shardRouting.replicaShards()) {
-                    status.addReplica(replicaShard, shutdown);
+                    status.addReplica(replicaShard, state, shutdown, includeDetails);
                 }
             }
         }
 
-        return createIndicator(status.getStatus(), status.getSummary(), status.getDetails(includeDetails), status.getImpacts());
+        return new HealthIndicatorResult(
+            name(),
+            component(),
+            status.getStatus(),
+            status.getSummary(),
+            status.getDetails(includeDetails),
+            status.getImpacts(),
+            status.getUserActions(includeDetails)
+        );
     }
 
-    private static class ShardAllocationCounts {
+    // TODO: #85572 Fill in help URLs once they are finalized
+    public static final UserAction.Definition ACTION_RESTORE_FROM_SNAPSHOT = new UserAction.Definition(
+        "restore_from_snapshot",
+        Explanations.Allocation.NO_COPIES,
+        null
+    );
+    public static final UserAction.Definition ACTION_ENABLE_ALLOCATIONS = new UserAction.Definition(
+        "enable_allocations",
+        "Elasticsearch isn't allowed to allocate shards from these indices because allocation for those shards has been disabled. "
+            + "Check that the ["
+            + EnableAllocationDecider.INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] index settings and the ["
+            + EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] cluster setting are not set to ["
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
+            + "].",
+        null
+    );
+    public static final UserAction.Definition ACTION_SHARD_LIMIT = new UserAction.Definition(
+        "increase_shard_limit",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because each node in the tier has "
+            + "reached its shard limit. Increase the values for the ["
+            + ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey()
+            + "] index setting on each index or add more nodes to the cluster.",
+        null
+    );
+    public static final UserAction.Definition ACTION_MIGRATE_TIERS = new UserAction.Definition(
+        "migrate_data_tiers",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because no nodes in the tier are "
+            + "compatible with the allocation filters in the index settings. Remove the conflicting allocation filters from each index's "
+            + "settings or try migrating to data tiers using the data tier migration action.",
```

Comment:
Should we add the actual API to run in the action message? `POST  /_ilm/migrate_to_data_tiers`

Reply:
Added in 3370adf